File: companion/backend/registry.py
"""BackendRegistry: 按 contact slug + source 选后端.

桌面 UI 的"蒸馏人格 / 真人"切换就是调 `get(slug, source)` 换实例。
每个 contact 可以同时拥有两个后端实例, 但两条对话历史存在不同的 sqlite 里,
不会互相污染。
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Literal

from .base import ChatBackend
from .distilled import DistilledBackend
from .feishu import FeishuBackend


Source = Literal["distilled", "feishu"]


class BackendRegistry:
    def __init__(self, companion_root: str | Path):
        self.root = Path(companion_root)
        self.personas_dir = self.root / "personas"
        self.feishu_storage = self.root / "runtime" / "feishu"
        self.contacts_index = self.root / "runtime" / "contacts.json"

        self._cache: dict[tuple[str, Source], ChatBackend] = {}

    def list_contacts(self) -> list[dict]:
        """列出所有联系人. 索引由 contacts.json 维护; 没有就扫 personas/."""
        if self.contacts_index.exists():
            return json.loads(self.contacts_index.read_text(encoding="utf-8"))

        out: list[dict] = []
        if self.personas_dir.exists():
            for d in sorted(self.personas_dir.iterdir()):
                meta = d / "meta.json"
                if meta.exists():
                    data = json.loads(meta.read_text(encoding="utf-8"))
                    out.append(
                        {
                            "slug": data.get("slug", d.name),
                            "name": data.get("name", d.name),
                            "has_distilled": True,
                            "has_feishu": False,
                            "feishu_chat_id": None,
                        }
                    )
        return out

    def register_feishu(self, slug: str, name: str, chat_id: str) -> None:
        """把一个 contact 的飞书 chat_id 记下来, 下次 get(slug, 'feishu') 用它."""
        contacts = self.list_contacts()
        found = False
        for c in contacts:
            if c["slug"] == slug:
                c["has_feishu"] = True
                c["feishu_chat_id"] = chat_id
                c["name"] = name
                found = True
                break
        if not found:
            contacts.append(
                {
                    "slug": slug,
                    "name": name,
                    "has_distilled": (self.personas_dir / slug / "SKILL.md").exists(),
                    "has_feishu": True,
                    "feishu_chat_id": chat_id,
                }
            )
        self.contacts_index.parent.mkdir(parents=True, exist_ok=True)
        self.contacts_index.write_text(json.dumps(contacts, ensure_ascii=False, indent=2), encoding="utf-8")

    def get(self, slug: str, source: Source) -> ChatBackend:
        key = (slug, source)
        if key in self._cache:
            return self._cache[key]

        if source == "distilled":
            backend: ChatBackend = DistilledBackend(self.personas_dir / slug)
        elif source == "feishu":
            contact = next((c for c in self.list_contacts() if c["slug"] == slug), None)
            if not contact or not contact.get("feishu_chat_id"):
                raise ValueError(f"联系人 {slug} 未绑定飞书 chat_id, 先调 register_feishu")
            backend = FeishuBackend(
                contact_slug=slug,
                chat_id=contact["feishu_chat_id"],
                display_name=contact.get("name", slug),
                storage_dir=self.feishu_storage,
            )
        else:
            raise ValueError(f"未知 source: {source}")

        self._cache[key] = backend
        return backend

File: packages/companion/src/backend/registry.ts
import fs from "node:fs";
import path from "node:path";
import type { Contact, Source } from "@petchat/shared";
import type { ChatBackend } from "./base.js";
import { DistilledBackend } from "./distilled.js";
import { FeishuBackend } from "./feishu.js";

/**
 * BackendRegistry resolves (slug, source) pairs into backend instances.
 * Two backends for the same slug keep history in separate sqlite files
 * so the distilled conversation and the real Feishu relay don't pollute
 * each other.
 */
export class BackendRegistry {
  readonly root: string;
  readonly personasDir: string;
  readonly feishuStorage: string;
  readonly contactsIndex: string;
  private cache = new Map<string, ChatBackend>();

  constructor(companionRoot: string) {
    this.root = companionRoot;
    this.personasDir = path.join(companionRoot, "personas");
    this.feishuStorage = path.join(companionRoot, "runtime", "feishu");
    this.contactsIndex = path.join(companionRoot, "runtime", "contacts.json");
  }

  listContacts(): Contact[] {
    if (fs.existsSync(this.contactsIndex)) {
      return JSON.parse(fs.readFileSync(this.contactsIndex, "utf-8")) as Contact[];
    }
    if (!fs.existsSync(this.personasDir)) return [];
    const out: Contact[] = [];
    for (const d of fs.readdirSync(this.personasDir).sort()) {
      const metaPath = path.join(this.personasDir, d, "meta.json");
      if (!fs.existsSync(metaPath)) continue;
      const meta = JSON.parse(fs.readFileSync(metaPath, "utf-8")) as {
        slug?: string;
        name?: string;
      };
      out.push({
        slug: meta.slug ?? d,
        name: meta.name ?? d,
        hasDistilled: true,
        hasFeishu: false,
        feishuChatId: null,
      });
    }
    return out;
  }

  registerFeishu(slug: string, name: string, chatId: string): void {
    const contacts = this.listContacts();
    const existing = contacts.find((c) => c.slug === slug);
    if (existing) {
      existing.hasFeishu = true;
      existing.feishuChatId = chatId;
      existing.name = name;
    } else {
      contacts.push({
        slug,
        name,
        hasDistilled: fs.existsSync(path.join(this.personasDir, slug, "SKILL.md")),
        hasFeishu: true,
        feishuChatId: chatId,
      });
    }
    fs.mkdirSync(path.dirname(this.contactsIndex), { recursive: true });
    fs.writeFileSync(
      this.contactsIndex,
      JSON.stringify(contacts, null, 2),
      "utf-8",
    );
  }

  get(slug: string, source: Source): ChatBackend {
    const key = `${slug}:${source}`;
    const cached = this.cache.get(key);
    if (cached) return cached;

    let backend: ChatBackend;
    if (source === "distilled") {
      backend = new DistilledBackend(path.join(this.personasDir, slug));
    } else {
      const contact = this.listContacts().find((c) => c.slug === slug);
      if (!contact || !contact.feishuChatId) {
        throw new Error(`联系人 ${slug} 未绑定飞书 chat_id, 先调用 registerFeishu`);
      }
      backend = new FeishuBackend({
        contactSlug: slug,
        chatId: contact.feishuChatId,
        displayName: contact.name ?? slug,
        storageDir: this.feishuStorage,
      });
    }
    this.cache.set(key, backend);
    return backend;
  }
}
